src/lib/storage/column_encoders/base_column_encoder.hpp

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include <boost/hana/type.hpp>
+
+#include <memory>
+
+#include "storage/base_value_column.hpp"
+#include "storage/encoded_columns/column_encoding_type.hpp"
+
+#include "all_type_variant.hpp"
+#include "resolve_type.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+namespace hana = boost::hana;
+
+/**
+ * @brief Base class of all column encoders
+ *
+ * Use the column_encoder.template.hpp to add new implementations!
+ */
+class BaseColumnEncoder {
+ public:
+  /**
+   * @brief Returns true if the encoder supports the given data type.
+   */
+  virtual bool supports(DataType data_type) const = 0;
+
+  /**
+   * @brief Encodes a value column with the given data type.
+   *
+   * @return encoded column if data type is supported else nullptr
+   */
+  virtual std::shared_ptr<BaseColumn> encode(DataType data_type, const std::shared_ptr<BaseValueColumn>& column) = 0;
+};
+
+template <typename Derived>
+class ColumnEncoder : public BaseColumnEncoder {
+ public:
+  /**
+   * @defgroup Virtual interface implementation
+   * @{
+   */
+  bool supports(DataType data_type) const final {
+    auto result = bool{};
+    resolve_data_type(data_type, [&](auto type_obj) { result = this->supports(type_obj); });
+    return result;
+  }
+
+  std::shared_ptr<BaseColumn> encode(DataType data_type, const std::shared_ptr<BaseValueColumn>& column) final {
```

Comment:
Comment? Something like "dispatches to the appropriate instantiation of "encode"?

Reply:
done